source/common/http/conn_pool_grid.cc

```diff
@@ -98,8 +115,13 @@ void ConnectivityGrid::WrapperCallbacks::onConnectionAttemptReady(
     ConnectionAttemptCallbacks* attempt, RequestEncoder& encoder,
     Upstream::HostDescriptionConstSharedPtr host, const StreamInfo::StreamInfo& info,
     absl::optional<Http::Protocol> protocol) {
+  ASSERT(host == grid_.host_);
   ENVOY_LOG(trace, "{} pool successfully connected to host '{}'.", describePool(attempt->pool()),
             grid_.host_->hostname());
+  if (http3_attempt_failed_) {
+    // If the HTTP/3 attempt failed but the other attempt succeeded, mark HTTP/3 as broken.
+    grid_.set_is_http3_broken(true);
```

Comment:
Don't we need to handle the case where H2 succeeds before H3 fails as well?

Reply:
Right, that's the TODO here:

  // TODO: Ensure that if HTTP/2 succeeds, we can allow the HTTP/3 connection to run to completion.                         

How 'bout I address this in a followup since I suspect that might result in a bit of refactoring?

Reply:
Turned out to be less work than I expected, so I've resolved this TODO and handle this case now. (Added a test for it)